src/main/groovy/com/morpheus/scvmm/sync/VirtualMachineSync.groovy

```diff
@@ -0,0 +1,490 @@
+            def volumeConfig = [
+                    name      : diskData.Name,
+                    size      : diskData.TotalSize?.toLong() ?: 0,
+                    rootVolume: diskData.VolumeType == 'BootAndSystem' || !server.volumes?.size(),
+                    //deviceName: (diskData.deviceName ?: provisionProvider.getDiskName(diskNumber)),
+                    deviceName: diskData.deviceName,
+                    externalId: diskData.ID,
+                    internalId: diskData.Name
+            ]
+            if (datastore)
+                volumeConfig.datastoreId = "${datastore.id}"
+            def storageVolume = buildStorageVolume(server.account ?: cloud.account, server, volumeConfig)
+            context.async.storageVolume.create([storageVolume], server).blockingGet()
+            maxStorage += storageVolume.maxStorage ?: 0l
+            diskNumber++
+            log.debug("added volume: ${storageVolume?.dump()}")
+        }
+    }
+
+    def updateMatchedStorageVolumes(updateItems, server, maxStorage, changes) {
+        updateItems?.each { updateMap ->
+            log.debug("updating volume: ${updateMap.masterItem}")
+            StorageVolume volume = updateMap.existingItem
+            def masterItem = updateMap.masterItem
+
+            def masterDiskSize = masterItem?.TotalSize?.toLong() ?: 0
+            def sizeRange = [min: (volume.maxStorage - ComputeUtility.ONE_GIGABYTE), max: (volume.maxStorage + ComputeUtility.ONE_GIGABYTE)]
+            def save = false
+            if (masterDiskSize && volume.maxStorage != masterDiskSize && (masterDiskSize <= sizeRange.min || masterDiskSize >= sizeRange.max)) {
+                volume.maxStorage = masterDiskSize
+                save = true
+            }
+            if (volume.internalId != masterItem.Name) {
+                volume.internalId = masterItem.Name
+                save = true
+            }
+            def isRootVolume = (masterItem?.VolumeType == 'BootAndSystem') || (server.volumes.size() == 1)
+            if (volume.rootVolume != isRootVolume) {
+                volume.rootVolume = isRootVolume
+                save = true
+            }
+            if (save) {
+                context.async.storageVolume.save(volume).blockingGet()
+                changes = true
+            }
+            maxStorage += masterDiskSize
+        }
+    }
+
+    def removeMissingStorageVolumes(removeItems, server, changes) {
+        removeItems?.each { currentVolume ->
+            log.debug "removing volume: ${currentVolume}"
+            changes = true
+            currentVolume.controller = null
+            currentVolume.datastore = null
+
+            context.async.storageVolume.save(currentVolume).blockingGet()
+            context.async.storageVolume.remove([currentVolume], server, changes).blockingGet()
+            context.async.storageVolume.remove(currentVolume).blockingGet()
```

Comment:
the list single remove looks like a duplicate of the bulk relmove on 428. the api call on 427 and 428 look like they could be converted into bulk calls.

Reply:
review comment changes added as suggested